# Unsupervised Learning: Clustering Lab





In [1]:
from sklearn.cluster import AgglomerativeClustering, KMeans
from sklearn.metrics import silhouette_score
import numpy as np
import pandas as pd
from scipy.io import arff

## 1. Initial practice with the K-means and HAC algorithms

### 1.1 (10%) K-means
Run K-means on this [Abalone Dataset.](https://raw.githubusercontent.com/cs472ta/CS472/master/datasets/abalone.arff)
The dataset was modified to be smaller. The last datapoint should be on line 359 or the point 0.585,0.46,0.185,0.922,0.3635,0.213,0.285,10. The remaining points are commented out. Treat the output class (last column) as an additional input feature. Create your K-Mmeans model with the paramaters K-means(n_clusters=3, init='random', n_init=1) 

Output the following:
- Class label for each point (labels_)
- The k=3 cluster centers (cluster_centers_)
- Number of iterations it took to converge (n_iter_)
- Total sum squared error of each point from its cluster center (inertia_)
- The total average silhouette score (see sklearn.metrics silhouette_score)

In [14]:
# K-means with Abalone
from sklearn.metrics import silhouette_score
import pandas as pd
from sklearn.cluster import KMeans


# get data
abaloneData = arff.loadarff('abalone.arff')

# put data into a data frame
abaloneDataFrame = pd.DataFrame(abaloneData[0])

clf = KMeans(n_clusters=3, init='random', n_init=1)

clf.fit(abaloneDataFrame)

print('Labels: ', clf.labels_, '\n')
print('Cluster centers: ', clf.cluster_centers_, '\n')
print('Number of Iterations to converge: ', clf.n_iter_, '\n')
print('Total sum squared error of each point from its cluser center: ', clf.inertia_, '\n')
print('Total average solhouette score: ', silhouette_score(abaloneDataFrame, clf.labels_), '\n')

Labels:  [0 2 2 1 2 2 0 0 2 0 1 1 1 1 1 1 2 1 2 2 1 1 1 2 1 1 1 1 0 1 1 0 0 0 1 2 0
 2 1 2 2 1 2 2 2 2 2 2 2 2 2 2 1 1 2 2 2 2 2 2 2 2 1 2 2 2 1 1 1 2 1 2 0 1
 1 0 2 1 1 1 2 1 0 0 1 1 1 1 2 1 0 1 1 1 0 1 2 2 2 1 2 0 0 1 1 1 1 1 2 2 2
 2 2 2 1 1 1 1 2 2 2 2 1 2 2 2 2 2 0 0 0 2 2 2 2 2 2 2 2 2 1 1 0 1 1 1 1 2
 2 2 0 2 1 1 1 1 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 2 2 2 2 2 2 2 2 1 1 0 1 1 1
 1 1 1 1 2 1 1 1 2 1 1 1 0 0 1] 

Cluster centers:  [[ 0.61366667  0.48933333  0.16716667  1.29283333  0.48815     0.25873333
   0.45488333 17.06666667]
 [ 0.56078652  0.44050562  0.15308989  0.94596067  0.37410112  0.20873596
   0.30108989 11.49438202]
 [ 0.40975309  0.31709877  0.10277778  0.39740741  0.16626543  0.09267284
   0.12505556  7.55555556]] 

Number of Iterations to converge:  4 

Total sum squared error of each point from its cluser center:  512.9856925785199 

Total average solhouette score:  0.5014855438589172 



**Discussion:**  

In the above cell, I ran a K-Means classifier with the parameters n_clusters=3, init='random', and n_init=1. The K-Means classifier, chooses n_clusters number of centroid points (in this case 3), and will group each point into the cluster of the nearest centroid point. Then, after grouping all points, the K-Means algorithm with recalculate the centroid point, then regroup all points accordingly. This process of calculating a centroid, then regrouping points into the group of the nearest centroid repeats until the centroids stop moving.  

After running the K-Means classifier on this dataset, I output the following information:  
1. Labels. First, I output the labels expected by the K-Means algorithm:  

[0 2 2 1 2 2 0 0 2 0 1 1 1 1 1 1 2 1 2 2 1 1 1 2 1 1 1 1 0 1 1 0 0 0 1 2 0
 2 1 2 2 1 2 2 2 2 2 2 2 2 2 2 1 1 2 2 2 2 2 2 2 2 1 2 2 2 1 1 1 2 1 2 0 1
 1 0 2 1 1 1 2 1 0 0 1 1 1 1 2 1 0 1 1 1 0 1 2 2 2 1 2 0 0 1 1 1 1 1 2 2 2
 2 2 2 1 1 1 1 2 2 2 2 1 2 2 2 2 2 0 0 0 2 2 2 2 2 2 2 2 2 1 1 0 1 1 1 1 2
 2 2 0 2 1 1 1 1 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 2 2 2 2 2 2 2 2 1 1 0 1 1 1
 1 1 1 1 2 1 1 1 2 1 1 1 0 0 1]  

This list is the cluster group (0, 1, or 2) for each point.  


2. Cluster Centers. The next thing I output from the K-Means algorithm is the Cluster Centers:  

[
    [ 0.61366667  0.48933333  0.16716667  1.29283333  0.48815     0.25873333 0.45488333 17.06666667],
    [ 0.56078652  0.44050562  0.15308989  0.94596067  0.37410112  0.20873596 0.30108989 11.49438202],
    [ 0.40975309  0.31709877  0.10277778  0.39740741  0.16626543  0.09267284 0.12505556  7.55555556]
]  

This output that you can see above provides the final coordinates of each cluster center. As expected, because there are 3 clusters, there are 3 centroid points output by the clf.cluster_centers_ attribute. 


3. Number of Iterations to converge: 4. This attribute, given by clf.n_iter_ tells us the number of times centroids were recalculated and instances regrouped before convergence (aka. centeroids stopped moving) occurred. In this case, it took 4 recalculations to converge.  


4. Total sum squared error of each point from its cluser center: 512.9856925785199. This value, given by clf.inertia_, is a measure of how close each point is to its cluster center. A small value here would mean that all points are very close to their final centroid point. A large value signals that each point is very far from its centroid point.  


5. Total average silhouette score:  0.5014855438589172. Silhouette score is an indication of 2 things. First, how closely grouped clusters are. And second, how distantly 

### 1.2 (10%) Hierarchical Agglomerative Clustering (HAC) 

Run HAC on the same [Abalone Dataset](https://raw.githubusercontent.com/cs472ta/CS472/master/datasets/abalone.arff) using complete linkage and k=3.

Output the following:
- Class label for each point (labels_)
- The total average silhouette score

In [ ]:
# HAC with Abalone

*Discussion*

## 2. K-means Clustering with the [Iris Dataset](https://raw.githubusercontent.com/cs472ta/CS472/master/datasets/iris.arff)
Use the Iris data set for 2.1 and 2.2.  Don't include the output label as one of the input features.

### 2.1 (20%) K-means Initial Centroids Experiments
K-means results differ based on the initial centroids used.
- Run K-means 5 times with *k*=4, each time with different initial random centroids (init="random) and with n_init=1.  Give inertia and silhouette scores for each run and discuss any variations in the results.
- SKlearn has a parameter that does this automatically (n_init).  n_init = z runs K-means z times, each with different random centroids and returns the clustering with the best SSE (intertia) of the z runs. Try it out and discuss how it does and how it compares with your 5 runs above.
- Sklearn also has a parameter (init:'K-means++') which runs a simpler fast version of K-means first on the data to come up with good initial centroids, and then runs regular K-means with this centroids.  Try it out (with n_init = 1) and discuss.

In [ ]:
# K-means initial centroid experiments

Results and Discussion

### 2.2 (20%) Silhouette Graphs
In this part you will show silhouette graphs for different *k* values.  Install the [Yellowbrick visualization package](https://www.scikit-yb.org/en/latest/quickstart.html) and import the [Silhouette Visualizer](https://www.scikit-yb.org/en/latest/api/cluster/silhouette.html).  This library includes lots of visualization packages which you might find useful. (Note: The YellowBrick silhouette visualizer does not currently support HAC).
- Show Silhouette graphs for clusterings with *k* = 2-6. Print the SSE (inertia) and total silhouette score for each.
- Learn with the default n_init = 10 to help insure a decent clustering.
- Using the silhouette graphs, choose which *k* you think is best and discuss why. Think about and discuss more than just the total silhouette score.

In [ ]:
# Iris Clustering with K-means and silhouette graphs
from yellowbrick.cluster import SilhouetteVisualizer

Discuss your results and justify which clustering is best based on the silhouette graphs

## 3 (20%) Iris Clustering with HAC

- Use the same dataset as above and learn with HAC clustering
- Create one table with silhouette scores for k=2-6 for each of the linkage options single, average, complete, and ward

In [ ]:
#HAC with Iris

*Discussion and linkage comparison*

## 4 (20%) Run both algorithms on a real world data
- Choose any real world data set which you have not used previously
- Use parameters of your choosing
- Try each algorithm a few times with different parameters and output one typical example of labels and silhouette scores for each algorithm
- Show the silhouette graph for at least one reasonable *k* value for K-means

In [ ]:
# Run both algoriths on a data set of your choice

*Discussion and comparison*

## 5. Extra Credit for Coding Your Own Clustering Algorithms
### 5.1 (Optional 10% extra credit) Code up the K-means clustering algorithm 
Below is a scaffold you could use if you want. As above, you only need to support numeric inputs, but think about how you would support nominal inputs and unknown values. Requirements for this task:
- Your model should support the methods shown in the example scaffold below.
- Ability to choose *k* and specify the *k* initial centroids.
- Run and show the cluster label for each point with both the Iris data set and the data set of your choice above.

### 5.2 (Optional 10% extra credit) Code up the HAC clustering algorithm 

- Your model should support the methods shown in the example scaffold below.
- HAC should support both single link and complete link options.
- HAC automatically generates all clusterings from *n* to 2.  You just need to output results for the curent chosen *k*.
- Run and show the cluster label for each point with both the Iris data set and the data set of your choice above.

Discussion and comparision of each model implemented

In [ ]:
from sklearn.base import BaseEstimator, ClassifierMixin, ClusterMixin

class KMEANSClustering(BaseEstimator,ClusterMixin):

    def __init__(self,k=3,debug=False): ## add parameters here
        """
        Args:
            k = how many final clusters to have
            debug = if debug is true use the first k instances as the initial centroids otherwise choose random points as the initial centroids.
        """
        self.k = k
        self.debug = debug

    def fit(self, X, y=None):
        """ Fit the data; In this lab this will make the K clusters :D
        Args:
            X (array-like): A 2D numpy array with the training data
            y (array-like): An optional argument. Clustering is usually unsupervised so you don't need labels
        Returns:
            self: this allows this to be chained, e.g. model.fit(X,y).predict(X_test)
        """
        return self
    
    def print_labels(self): # Print the cluster label for each data point
        pass

In [ ]:
class HACClustering(BaseEstimator,ClusterMixin):

    def __init__(self,k=3,link_type='single'): ## add parameters here
        """
        Args:
            k = how many final clusters to have
            link_type = single or complete. when combining two clusters use complete link or single link
        """
        self.link_type = link_type
        self.k = k
        
    def fit(self, X, y=None):
        """ Fit the data; In this lab this will make the K clusters :D
        Args:
            X (array-like): A 2D numpy array with the training data
            y (array-like): An optional argument. Clustering is usually unsupervised so you don't need labels
        Returns:
            self: this allows this to be chained, e.g. model.fit(X,y).predict(X_test)
        """
        return self
    
    def print_labels(self): # Print the cluster label for each data point
        pass